Add mask identifier for multiplexed src_mask/src_key_padding_mask in BT

nmacchioni opened the pull request:
Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/81947

Transformer fastpath multiplexes two arguments, src_mask [seq_len x seq_len] and src_key_padding_mask [batch_size x seq_len], and later deduces the type based on mask shape.

In the event that batch_size == seq_len, any src_mask is wrongly interpreted as a src_key padding_mask. This is fixed by requiring a mask_type identifier be supplied whenever batch_size == seq_len.

Additionally, added support for src_mask in masked_softmax CPU path.

Test Plan: existing unit tests + new unit tests (batch_size == seq_len)

Differential Revision: D37932240

zrphercule reviewed (commented):
Overall LGTM, comments in below

zrphercule commented on aten/src/ATen/native/cuda/SoftMax.cu:
```diff
@@ -957,18 +957,19 @@ TORCH_IMPL_FUNC(softmax_backward_cuda_out)
   // If input is [B, H, T, T] and mask is [B, T]
   // we have special fast kernel
-  bool is_BxT_mask = (input_.dim() == 4 && mask_.dim() == 2 && input_.size(0) == mask_.size(0) && input_.size(2) == mask_.size(1) && input_.size(3) == mask_.size(1));
+  bool is_BxT_mask = (mask_type_ == 2) && (input_.dim() == 4 && mask_.dim() == 2 && input_.size(0) == mask_.size(0) && input_.size(2) == mask_.size(1) && input_.size(3) == mask_.size(1));
```
Can we use Enum instead of hardcoded number here? Or at least we need some comments indicate what mask_type==1 or ==2 means

nmacchioni replied:
I was originally looking to use an Enum but couldn't figure out where the best location for it was? If you had a suggestion that would be appreciated and then I could go back and change the types.

zrphercule replied:
> 
Since this is being passed as param of function, I guess you can keep it as number. But please put some comments at the beginning of the function, to clarify each number's meaning.

zrphercule commented on aten/src/ATen/native/SoftMax.cpp:
```diff
@@ -131,7 +131,10 @@ void host_softmax(
-    bool* mask = nullptr) {
+    bool* mask = nullptr,
+    const c10::optional<int64_t> mask_type = NULL) {
+  TORCH_CHECK(!mask || mask_type,
```
Please make sure adding mask_type for all existing code

nmacchioni commented:
@zrphercule Could you please approve the above workflows

zrphercule reviewed (approved):
Thanks!

nmacchioni committed:
Add mask identifier for multiplexed src_mask/src_key_padding_mask in BT (#81947)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/81947

Transformer fastpath multiplexes two arguments, src_mask [seq_len x seq_len] and src_key_padding_mask [batch_size x seq_len], and later deduces the type based on mask shape.

In the event that batch_size == seq_len, any src_mask is wrongly interpreted as a src_key padding_mask. This is fixed by requiring a mask_type identifier be supplied whenever batch_size == seq_len.

TODO: Add support for src_mask in masked_softmax CPU path.

Test Plan: existing unit tests + new unit tests (batch_size == seq_len)

Reviewed By: mikekgfb

Differential Revision: D37932240

fbshipit-source-id: 3ebce90095c97975fba8018e3e7e27b4f24c49ca

zrphercule commented on aten/src/ATen/native/SoftMax.cpp:
```diff
@@ -131,7 +131,18 @@ void host_softmax(
+  if (MaskedSoftMax) {
+    TORCH_CHECK(mask_type_.has_value(), "Mask Type should be defined");
+    int64_t mask_type = mask_type_.value();
+    TORCH_CHECK((mask_type == 0) || (mask_type == 1), "Mask Type should be 0 (src_mask) or 1 (src_key_padding_mask)");
```
Maybe re-org this part of logic a bit? since mask_type == 0 does not work on CPU

nmacchioni replied:
Since adding support for mask_type == 0 is in progress I chose to have that check separate as it will be removed soon

zrphercule commented on test/test_nn.py:
```diff
@@ -20862,6 +20867,17 @@ def test_multihead_attn_in_proj_weight_none(self, device, dtype):
         key = torch.rand(4, 4, 2, dtype=dtype, device=device)
         mha(query, key, key)
 
+    @onlyCPU
```
Why only CPU?

nmacchioni replied:
Test is specifically targeted at addressing issue #81129 which only occurs on CPU

zrphercule reviewed (commented):
I feel this diff itself overall LGTM, but will it break internal CI? Since you changed some API

zrphercule reviewed (approved):
OK CI looks fine, thanks

nmacchioni commented:
@pytorchbot merge
